eigensync/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+            peer,
+            request_id,
+            error,
+        })) => {
+            let sender = match response_map.remove(&request_id) {
+                Some(sender) => sender,
+                None => {
+                    tracing::error!("No sender for request id {:?}", request_id);
+                    return Ok(());
+                }
+            };
+
+            let _ = sender.send(Err(error.to_string()));
+        }
+        SwarmEvent::ConnectionEstablished { peer_id: _peer_id, .. } => {
+            // send the connection established signal
+            if let Some(sender) = connection_established.take() {
+                let _ = sender.send(());
+            }
+        },
+        other => tracing::error!("Received event: {:?}", other),
+    })
+}
+
+pub struct EigensyncHandle<T: Reconcile + Hydrate + Default + Debug> {
+    pub document: AutoCommit,
+    sender: UnboundedSender<ChannelRequest>,
+    encryption_key: [u8; 32],
+    _marker: PhantomData<T>,
+}
+
+impl<T: Reconcile + Hydrate + Default + Debug> EigensyncHandle<T> {
+    pub async fn new(server_addr: Multiaddr, server_id: PeerId, keypair: identity::Keypair, encryption_key: [u8; 32]) -> anyhow::Result<Self> {
+
+        let mut swarm = SwarmBuilder::with_existing_identity(keypair)
+            .with_tokio()
+            .with_tcp(
+                tcp::Config::default(),
+                noise::Config::new,
+                yamux::Config::default,
+            )
+            .context("Failed to create TCP transport")?
+            .with_behaviour(|_| Ok(client()))
+            .context("Failed to create behaviour")?
+            .with_swarm_config(|cfg| cfg.with_idle_connection_timeout(Duration::MAX))
+            .build();
+                
+        swarm.add_peer_address(server_id.clone(), server_addr.clone());
+
+        swarm.dial(server_id).context("Failed to dial")?;
+        
+        let (sender, receiver) = unbounded_channel();
+        let (connection_ready_tx, connection_ready_rx) = oneshot::channel();
+
+        task::spawn(async move {
+            SyncLoop::new(receiver, swarm, connection_ready_tx).await.run(server_id).await.unwrap();
+        });
+
+        connection_ready_rx.await.context("Failed to establish connection")?;
+
+        let document = AutoCommit::new().with_actor(ActorId::random());
+
+        let mut handle = Self {
+            document,
+            _marker: PhantomData,
+            sender,
+            encryption_key: encryption_key.clone(),
+        };
+
+        // Initial pull from server. If it fails, continue (we may be offline).
+        if let Err(e) = handle.sync_with_server().await {
+            tracing::error!("Initial eigensync pull failed: {:?}", e);
+        }
+
+        // Seed default only if the document is still empty after the pull.
+        if handle.document.get_changes(&[]).is_empty() {
+            let state = T::default();
```

Comment:
This means that changing the default value of a type may break compatibility with the automerge history. That's not good. We will have to come up with a better solution to this.